### PR TITLE
Enable exp10 libcall on linux

### PR DESCRIPTION
**Pull request by Krishna-13-cyber**

### CODE REVIEW

**Comment by arsenm** [comment]
@RKSimon should exp10/exp10f be available for -ps4/-ps5 triples?

File: llvm/test/Transforms/InstCombine/pow-1.ll [hunk 1]
```diff
@@ -1,3 +1,4 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/pow-1.ll:1]: Switch to generated checks should be committed separately

File: llvm/lib/Analysis/TargetLibraryInfo.cpp [hunk 2]
```diff
@@ -562,12 +562,9 @@ static void initialize(TargetLibraryInfoImpl &TLI, const Triple &T,
...
     // on Linux.
     //
-    // Fall through to disable all of them.
-    [[fallthrough]];
-  default:
-    TLI.setUnavailable(LibFunc_exp10);
-    TLI.setUnavailable(LibFunc_exp10f);
-    TLI.setUnavailable(LibFunc_exp10l);
+    TLI.setAvailableWithName(LibFunc_exp10, "__exp10");
+    TLI.setAvailableWithName(LibFunc_exp10f, "__exp10f");
```
Comment (pogo59) [inline llvm/lib/Analysis/TargetLibraryInfo.cpp:566]: exp10l?

Is this a fully-covered switch? (no longer has a default case)

File: llvm/test/Transforms/InstCombine/pow-exp.ll [hunk 3]
```diff
@@ -139,9 +139,10 @@ define fp128 @powl_exp2l_not_fast(fp128 %x, fp128 %y) {
 ; TODO: exp10() is not widely enabled by many targets yet.
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/pow-exp.ll:139]: Update todo?

File: llvm/lib/Analysis/TargetLibraryInfo.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:566]: This comment needs updating too

File: llvm/test/Transforms/InstCombine/pow-exp.ll (hunk 3 above)
Comment (Krishna-13-cyber) [reply llvm/test/Transforms/InstCombine/pow-exp.ll]: Updated.

File: llvm/lib/Analysis/TargetLibraryInfo.cpp (hunk 2 above)
Comment (Krishna-13-cyber) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:566]: Updated, no regressions present.

Comment (arsenm) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:566]: The comment hasn't been updated?

**Review by arsenm — CHANGES_REQUESTED** [review]

File: llvm/lib/Analysis/TargetLibraryInfo.cpp (hunk 2 above)
Comment (Krishna-13-cyber) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:566]: Yes, I can see the exp10l update in *Files changed*

Comment (Krishna-13-cyber) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:561]: Ah apologies, have rectified it. Thanks!

File: llvm/lib/Analysis/TargetLibraryInfo.cpp [hunk 4]
```diff
@@ -557,11 +557,6 @@ static void initialize(TargetLibraryInfoImpl &TLI, const Triple &T,
     }
     break;
   case Triple::Linux:
-    // exp10, exp10f, exp10l is available on Linux (GLIBC) but are extremely
-    // buggy prior to glibc version 2.18. Until this version is widely deployed
-    // or we have a reasonable detection strategy, we cannot use exp10 reliably
-    // on Linux.
-    //
```
Comment (arsenm) [inline llvm/lib/Analysis/TargetLibraryInfo.cpp:564]: I would leave some kind of historical note

Comment (Krishna-13-cyber) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:564]: Oh okay, thanks again!

Comment (arsenm) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:564]: Just update the "Until this version is widely deployed or we have a reasonable detection strategy, we cannot use exp10 reliably" part to something about how this is so old it doesn't matter

Comment (Krishna-13-cyber) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:564]: "As this version is so old, we don't really need to worry about using exp10 until we find more good ways to detect it."

Does this look fine?

Comment (arsenm) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:564]: The "until we find more good ways to detect it." reads off. Can drop it or rephrase

Comment (Krishna-13-cyber) [reply llvm/lib/Analysis/TargetLibraryInfo.cpp:564]: Yes, I will drop this.

File: llvm/lib/Analysis/TargetLibraryInfo.cpp [hunk 5]
```diff
@@ -557,17 +557,12 @@ static void initialize(TargetLibraryInfoImpl &TLI, const Triple &T,
     }
     break;
   case Triple::Linux:
-    // exp10, exp10f, exp10l is available on Linux (GLIBC) but are extremely
-    // buggy prior to glibc version 2.18. Until this version is widely deployed
```
Comment (arsenm) [inline llvm/lib/Analysis/TargetLibraryInfo.cpp:561]: You dropped the part that mentioned the version

**Review by arsenm — APPROVED** [review]

**Comment by arsenm** [comment]
Bot failure is real

File: llvm/lib/Analysis/TargetLibraryInfo.cpp [hunk 6]
```diff
@@ -558,16 +558,12 @@ static void initialize(TargetLibraryInfoImpl &TLI, const Triple &T,
     break;
   case Triple::Linux:
     // exp10, exp10f, exp10l is available on Linux (GLIBC) but are extremely
-    // buggy prior to glibc version 2.18. Until this version is widely deployed
-    // or we have a reasonable detection strategy, we cannot use exp10 reliably
-    // on Linux.
-    //
-    // Fall through to disable all of them.
-    [[fallthrough]];
-  default:
```
Comment (mstorsjo) [inline llvm/lib/Analysis/TargetLibraryInfo.cpp:567]: This change here caused a quite noisy warning from GCC:
```
../lib/Analysis/TargetLibraryInfo.cpp: In function ‘void initialize(llvm::TargetLibraryInfoImpl&, const llvm::Triple&, llvm::ArrayRef<llvm::StringLiteral>)’:
../lib/Analysis/TargetLibraryInfo.cpp:533:10: warning: enumeration value ‘UnknownOS’ not handled in switch [-Wswitch]
  533 |   switch (T.getOS()) {
      |          ^
../lib/Analysis/TargetLibraryInfo.cpp:533:10: warning: enumeration value ‘Darwin’ not handled in switch [-Wswitch]
../lib/Analysis/TargetLibraryInfo.cpp:533:10: warning: enumeration value ‘DragonFly’ not handled in switch [-Wswitch]
../lib/Analysis/TargetLibraryInfo.cpp:533:10: warning: enumeration value ‘FreeBSD’ not handled in switch [-Wswitch]
../lib/Analysis/TargetLibraryInfo.cpp:533:10: warning: enumeration value ‘Fuchsia’ not handled in switch [-Wswitch]
../lib/Analysis/TargetLibraryInfo.cpp:533:10: warning: enumeration value ‘KFreeBSD’ not handled in switch [-Wswitch]
```
(and so on for all OSes known by `Triple`)

So ideally we'd keep the default case to silence this compiler warning (or not enable `-Wswitch`, wherever that comes from).

**Comment by ecnelises** [comment]
Should this patch has nothing to do with PS? I tried adding `default:` label back, but `double-float-shrink-1.ll` `pow-1.ll` `pow-exp.ll` fail. If I understand this path correctly, cases other than linux should not be affected.

**Comment by nikic** [comment]
I went ahead and reverted this change. I could have fixed the uncovered switch warning, but the change in general looks problematic to me.

The comment indicates that these functions are broken prior to glibc 2.18 -- however, a common glibc baseline for binary distribution is glibc 2.17, because it is used by CentOS/RHEL 7. This is the glibc baseline that Rust uses, for example. If you want to increase the glibc requirements of LLVM-generated binaries, then that needs a wider discussion than this pull request (or more research to show that this is safe, e.g. because distros using glibc 2.17 have backported the necessary fixes or so).

There are also some buildbots failing with `undefined symbol: __exp10` after this change, for example https://lab.llvm.org/buildbot/#/builders/239/builds/6232.

**Pull request merged**